src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1043,37 +1059,106 @@ export class BundleDataClient {
-                  matchedDeposit.inputToken,
-                  matchedDeposit.originChainId,
-                  matchedDeposit.outputToken,
-                  matchedDeposit.destinationChainId,
-                  matchedDeposit.quoteBlockNumber
-                ) ||
-                // Deposit must not have expired in this bundle.
-                slowFillRequest.fillDeadline < bundleBlockTimestamps[destinationChainId][1]
-              ) {
-                // TODO: Invalid slow fill request. Maybe worth logging.
-                return;
-              }
               validatedBundleSlowFills.push(matchedDeposit);
             }
           }
         );
 
+        // The above loops for adding deposits, fills, and then slow fill requests to the relay hash dictionary
+        // ignore any events that are after the bundle block range. Because of that, we should consider that there
+        // are deposits in this bundle that correspond to fills that were sent in a prior bundle that have not
+        // yet been refunded. These fills are also known as "pre-fills" from here on.
+        const originBlockRange = getBlockRangeForChain(blockRangesForChains, originChainId, chainIds);
+
+        await mapAsync(
+          Object.values(v3RelayHashes).filter(
+            ({ deposit }) =>
+              deposit &&
+              deposit.originChainId === originChainId &&
+              deposit.destinationChainId === destinationChainId &&
+              deposit.blockNumber >= originBlockRange[0] &&
+              deposit.blockNumber <= originBlockRange[1] &&
+              !isZeroValueDeposit(deposit)
+          ),
+          async ({ deposit, fill, slowFillRequest }) => {
+            if (!deposit) throw new Error("Deposit should exist in relay hash dictionary.");
+            // We don't check the deposit's fillDeadline here because we are ok if the deposit expires in this bundle
+            // and we issue an expiry refund for it. This expired deposit could also have been pre-filled and we just
+            // want to make sure in this code block that all valid pre-fills get refunded once the deposit appears.
+            // If a pre-fill gets refunded and its deposit expired and gets refunded as well, then there is no net loss
+            // to the protocol.
+
+            // If fill exists in memory, then the only case in which we need to create a refund is if the
+            // the fill occurred in a previous bundle.
+            if (fill) {
+              if (!isSlowFill(fill) && fill.blockNumber < destinationChainBlockRange[0]) {
+                // If fill is in the current bundle then we can assume there is already a refund for it, so only
+                // include this pre fill if the fill is in an older bundle. If fill is after this current bundle, then
+                // we won't consider it, following the previous treatment of fills after the bundle block range.
+                validatedBundleV3Fills.push({
```

Comment:
I think this algorithm makes sense. Do you think it is worthwhile to add a log here, at least temporarily, so that we can verify that we aren't double counting deposits (as a sanity check). We would also be able to do a pre-fill and make sure it is included by this code block.

Reply:
I added unit tests here to check against this case: https://github.com/across-protocol/relayer/pull/2010 which I prefer over logs. Do you think duplicate refunds is a big risk based on the implementation?

Reply:
No not really. I can't think of a case where we would duplicate refund, but I just suggested it in case I am missing one. We can just alternatively run this locally in parallel with the "older" dataworker implementation and verify that the roots match.

Reply:
I was able to add a unit test here: b099cf0e6f910152b60c7dffd0618eae25a039fa

note that this test depends now on these SDK changes to support duplicate deposits https://github.com/across-protocol/sdk/pull/838